relayer/chains/evm/listener.go

```diff
@@ -81,120 +70,82 @@ func (p *Provider) Listener(ctx context.Context, lastSavedHeight uint64, blockIn
-						zap.Int64("gas-used", int64(lbn.Header.GasUsed)))
-
-					messages, err := p.FindMessages(ctx, lbn)
-					if err != nil {
-						return errors.Wrapf(err, "receiveLoop: callback: %v", err)
-					}
-					blockInfoChan <- &relayertypes.BlockInfo{
-						Height:   lbn.Height.Uint64(),
-						Messages: messages,
-					}
-				}
 
-				if lbn, bn = bn, nil; len(bnch) > 0 {
-					bn = <-bnch
-				}
-			}
-			// remove unprocessed notifications
-			for len(bnch) > 0 {
-				<-bnch
-			}
-
-		default:
-			if next >= latest {
-				continue
+			if isSubError {
+				startHeight = p.GetLastSavedBlockHeight()
 			}
 
-			qch := make(chan *bnq, cap(bnch))
-			for i := next; i < latest && len(qch) < cap(qch); i++ {
-				qch <- &bnq{i, nil, nil, 3} // fill bch with requests
+			var blockReqs []*blockReq
+			for i := startHeight; i < latestHeight; i += p.cfg.BlockBatchSize {
+				end := i + p.cfg.BlockBatchSize
+				if end > latestHeight {
+					end = latestHeight
+				}
+				blockReqs = append(blockReqs, &blockReq{start: i, end: end, retry: maxBlockQueryFailedRetry})
 			}
-			bns := make([]*types.BlockNotification, 0, len(qch))
-			for q := range qch {
-				switch {
-				case q.err != nil:
-					if q.retry > 0 {
-						if !strings.HasSuffix(q.err.Error(), "requested block number greater than current block number") {
-							q.retry--
-							q.v, q.err = nil, nil
-							qch <- q
-							continue
-						}
-						if latest >= q.h {
-							latest = q.h - 1
+			totalReqs := len(blockReqs)
+			// Calculate the size of each chunk
+			chunkSize := (totalReqs + concurrency - 1) / concurrency
+
+			var wg sync.WaitGroup
+
+			for i := 0; i < totalReqs; i += chunkSize {
+				wg.Add(1)
+
+				go func(blockReqsChunk []*blockReq, wg *sync.WaitGroup) {
+					defer wg.Done()
+					for _, br := range blockReqsChunk {
+						filter := ethereum.FilterQuery{
+							FromBlock: new(big.Int).SetUint64(br.start),
+							ToBlock:   new(big.Int).SetUint64(br.end),
+							Addresses: p.blockReq.Addresses,
+							Topics:    p.blockReq.Topics,
 						}
-					}
-					// r.Log.Debugf("receiveLoop: bnq: h=%d:%v, %v", q.h, q.v.Header.Hash(), q.err)
-					bns = append(bns, nil)
-					if len(bns) == cap(bns) {
-						close(qch)
-					}
-
-				case q.v != nil:
-					bns = append(bns, q.v)
-					if len(bns) == cap(bns) {
-						close(qch)
-					}
-				default:
-					go func(q *bnq) {
-						defer func() {
-							qch <- q
-						}()
-						if q.v == nil {
-							q.v = new(types.BlockNotification)
-						}
-						q.v.Height = new(big.Int).SetUint64(q.h)
-						q.v.Header, q.err = p.client.GetHeaderByHeight(ctx, q.v.Height)
-						if q.err != nil {
-							q.err = errors.Wrapf(q.err, "GetEvmHeaderByHeight %v", q.err)
-							return
+						logs, err := p.getLogsRetry(ctx, filter, br.retry)
+						if err != nil {
+							p.log.Warn("failed to fetch blocks", zap.Uint64("from", br.start), zap.Uint64("to", br.end), zap.Error(err))
```

Comment:
After this failure, what happens to set of blocks that we queried for event logs. We are just ignoring those blocks. May be we should have better strategy. Shouldn't we?

Reply:
Yep, This is just to mitigate the rpc failures like rate limits. 
I don't think this call would fail other than rpc failures.

Reply:
Yes dai, irrespective of type of failures, we should  not drop the set of block queried so easily. May be we should track in our db persistently, and retry later after fixing the issue. The concern is, if there is a event that has to delivered within this set of blocks then we just missed it, right?